feat(new-trace): Adding trace drawer.

Abdkhan14 opened the pull request:
<img width="1230" alt="Screenshot 2024-03-05 at 12 43 02 PM" src="https://github.com/getsentry/sentry/assets/60121741/59b3780f-aac1-4dfc-a6e1-a190c77ed2ff">
<img width="1221" alt="Screenshot 2024-03-05 at 12 43 16 PM" src="https://github.com/getsentry/sentry/assets/60121741/042b15b5-9a82-46fa-8ec9-1234b49c1036">
<img width="1221" alt="Screenshot 2024-03-05 at 12 43 42 PM" src="https://github.com/getsentry/sentry/assets/60121741/9a252e6f-53ec-499f-94e5-62cfdac98182">
<img width="1224" alt="Screenshot 2024-03-05 at 12 46 26 PM" src="https://github.com/getsentry/sentry/assets/60121741/b2548e0b-bd7a-4172-8300-abfed66befdf">

Commit:
feat(new-trace): Adding details for all trace row types.

Commit:
feat(new-trace): Updated trace error children count.

Abdkhan14 committed:
feat(new-trace): Added static bottom detail panel. (#66416)

https://github.com/getsentry/sentry/assets/60121741/9d757d1b-e318-4b87-8d8a-9d91dfef7e0b

---------

Co-authored-by: Abdullah Khan <abdullahkhan@PG9Y57YDXQ.local>

JonasBa commented on static/app/views/performance/newTraceDetails/trace.tsx:
```diff
@@ -322,10 +321,42 @@ function Trace({
+        return;
+      }
+
+      const traceViewElement: HTMLDivElement | null =
```
```suggestion
      const traceElement: HTMLDivElement | null =
```

JonasBa replied:
Minor, but you should just use HTMLElement types here since the tagname doesnt really matter. If we were to ever change the container type, you wouldnt need to update the type signatures

JonasBa commented on static/app/views/performance/newTraceDetails/trace.tsx:
```diff
@@ -322,10 +321,42 @@ function Trace({
+
+      const traceViewElement: HTMLDivElement | null =
+        'current' in containerRef ? containerRef.current : containerRef;
+      const detailPanelElement: HTMLDivElement | null =
```
```suggestion
      const detailElement: HTMLDivElement | null =
```

JonasBa commented on static/app/views/performance/newTraceDetails/trace.tsx:
```diff
@@ -322,10 +321,42 @@ function Trace({
+      const clickedTraceView =
+        !traceViewElement || traceViewElement.contains(event.target as Node);
+      const clickedDetailPanel = detailPanelElement?.contains(event.target as Node);
+      if (clickedTraceView || clickedDetailPanel) {
```
cant this just be 
```suggestion
      if (traceViewElement?.contains?.(event.target) || detailElement?.contains?.(event.target) {
```
Why do we ned to care about if the panel element is there or not?

Abdkhan14 replied:
Otherwise we lose the row highlighting, lets say during resizing. We don't want to treat the drawer as an outside click.

JonasBa commented on static/app/views/performance/newTraceDetails/traceDrawer/details/errorNodeDetails.tsx:
```diff
@@ -0,0 +1,105 @@
+
+import type {TraceTree, TraceTreeNode} from '../../traceTree';
+
+export default function ErrorNodeDetails({
```
lets stick to named exports
```suggestion
export function ErrorNodeDetails({
```

JonasBa commented on static/app/views/performance/newTraceDetails/traceDrawer/details/missingInstrumentaionNodeDetails.tsx:
```diff
@@ -0,0 +1,74 @@
+  );
+}
+
+const Wrapper = styled('div')`
```
These are being replicated in every component... Can you share these instead?

JonasBa commented on static/app/views/performance/newTraceDetails/traceDrawer/tabs/nodeDetails.tsx:
```diff
@@ -0,0 +1,59 @@
+import styled from '@emotion/styled';
```
`static/app/views/performance/newTraceDetails/traceDrawer/tabs/details.tsx`

JonasBa commented on static/app/views/performance/newTraceDetails/traceDrawer/tabs/nodeDetails.tsx:
```diff
@@ -0,0 +1,59 @@
+import SpanNodeDetails from '../details/spanNodeDetails';
+import TransactionNodeDetails from '../details/transactionNodeDetails';
+
+export default function NodeDetail({
```
named exports

JonasBa commented on static/app/views/performance/newTraceDetails/traceDrawer/tabs/nodeDetails.tsx:
```diff
@@ -0,0 +1,59 @@
+    <SiblingAutogroupNodeDetails node={node} />
+  ) : isMissingInstrumentationNode(node) ? (
+    <MissingInstrumentationNodeDetails node={node} />
+  ) : null;
```
This needs to throw if we cant render as all clickable nodes need to be able to render their data

JonasBa commented on static/app/views/performance/newTraceDetails/traceDrawer/tabs/nodeDetails.tsx:
```diff
@@ -0,0 +1,59 @@
+  organization: Organization;
+}) {
+  if (!node) {
+    return <NoDetail>{t('Click on a row in the trace view for its details')}</NoDetail>;
```
This should not be possible - if a user clicks away, then tab needs to switch to the trace tab as opposed to displaying an empty view

JonasBa commented on static/app/views/performance/newTraceDetails/traceDrawer/tabs/nodeDetails.tsx:
```diff
@@ -0,0 +1,59 @@
+  location,
+}: {
+  location: Location;
+  node: TraceTreeNode<TraceTree.NodeValue> | null;
```
```suggestion
  node: TraceTreeNode<TraceTree.NodeValue>;
```

JonasBa commented on static/app/views/performance/newTraceDetails/traceDrawer/tabs/traceLevelDetails.tsx:
```diff
@@ -21,7 +21,7 @@ import type {UseApiQueryResult} from 'sentry/utils/queryClient';
 import type RequestError from 'sentry/utils/requestError/requestError';
```
`static/app/views/performance/newTraceDetails/traceDrawer/tabs/trace.tsx`

JonasBa commented on static/app/views/performance/newTraceDetails/traceDrawer/traceDrawer.tsx:
```diff
@@ -0,0 +1,201 @@
+  };
+
+  return (
+    <PanelWrapper size={size} ref={panelRef}>
```
```suggestion
    <PanelWrapper size={size} ref={ref => props.setDetailPanelRef(ref)}>
```

JonasBa commented on static/app/views/performance/newTraceDetails/traceDrawer/traceDrawer.tsx:
```diff
@@ -0,0 +1,201 @@
+    };
+  }, [size, isResizing]);
+
+  useEffect(() => {
+    props.setDetailPanelRef(panelRef);
+  }, [panelRef, props]);
```
refs do not trigger rerenders
```suggestion
```

JonasBa commented on static/app/views/performance/newTraceDetails/traceDrawer/traceDrawer.tsx:
```diff
@@ -0,0 +1,201 @@
+    props.setDetailPanelRef(panelRef);
+  }, [panelRef, props]);
+
+  const handleMouseDown = e => {
```
```suggestion
  const handleMouseDown = useCallback(e => {
```

JonasBa commented on static/app/views/performance/newTraceDetails/traceDrawer/traceDrawer.tsx:
```diff
@@ -0,0 +1,201 @@
+
+  const handleMouseDown = e => {
+    e.preventDefault();
+    setIsResizing(true);
```
You dont need to rerender here, this can be a ref

JonasBa commented on static/app/views/performance/newTraceDetails/traceDrawer/traceDrawer.tsx:
```diff
@@ -0,0 +1,201 @@
+      </TabsContainer>
+
+      <Content>
+        {props.activeTab === 'trace_data' && (
```
Content rendering needs to be exclusive, so check for props.activeTab === trace ? <Trace/> : props.activeTab === "node"

JonasBa commented on static/app/views/performance/newTraceDetails/traceTree.tsx:
```diff
@@ -121,7 +121,7 @@ export declare namespace TraceTree {
   }
 
-  interface ChildrenAutogroup {
+  interface ChildrenAutogroup extends RawSpanType {
```
Why?

Abdkhan14 replied:
It's true, we spread the head node's value when we initialize the Auto-Grouped node [link](https://github.com/getsentry/sentry/blob/master/static/app/views/performance/newTraceDetails/traceTree.tsx#L515)

Abdkhan14 commented on static/app/views/performance/newTraceDetails/index.tsx:
```diff
@@ -224,20 +225,16 @@ function TraceViewContent(props: TraceViewContentProps) {
-      setDetailNode(prevNode => {
-        return prevNode === node ? null : node;
-      });
+      setActiveTab(node && !isTraceNode(node) ? 'node_detail' : 'trace_data');
```
We also want to go to the Trace tab when there isn't a clicked node.

Abdkhan14 commented on static/app/views/performance/newTraceDetails/trace.tsx:
```diff
@@ -322,10 +321,42 @@ function Trace({
       query: queryParamsWithoutNode,
     });
+
+    setDetailNode(null);
```
Clicking outside should get rid of the detail tab from the drawer.

Commit:
feat(new-trace): Addressing minor pr suggestions.

Commit:
feat(new-trace): Getting rid of resizing state.

JonasBa committed:
ref(trace): update components

JonasBa committed:
ref(trace): update rendering

JonasBa committed:
ref(trace): add scroll to button

Commit:
feat(new-trace): Resolving  conflicts.

getsantry[bot] committed:
:hammer_and_wrench: apply pre-commit fixes

Abdkhan14 committed:
Merge branch 'master' into abdk/trace-details-for-all-rows

Commit:
feat(new-trace): Reverting changes to api.tsx

Abdkhan14 committed:
feat(new-trace): Updated span description to include query highlighting. (#66709)

<img width="1249" alt="Screenshot 2024-03-11 at 2 16 48 PM"
src="https://github.com/getsentry/sentry/assets/60121741/44165007-ef62-40ef-a6b8-131cfee62b19">

---------

Co-authored-by: Abdullah Khan <abdullahkhan@PG9Y57YDXQ.local>

Commit:
feat(new-trace): Reverting teamSelector file change

Commit:
feat(new-trace): merging conflicts
